cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/distributed/k8s/SparkContainerDriverLauncher.java

```
@@ -0,0 +1,258 @@
+    ArtifactLocalizerClient fetchArtifacts = createArtifactLocalizerClient(cConf, hConf);
+
+    //Create plugin location for storing plugin jars
+    Path pluginsLocation = new File(WORKING_DIRECTORY).getAbsoluteFile().toPath().resolve("artifacts_archive.jar")
+      .toAbsolutePath();
+    Files.createDirectories(pluginsLocation);
+
+    // Fetching plugin artifacts from app-fabric
+    for (Plugin plugin : spec.getPlugins().values()) {
+      File tempLocation = fetchArtifacts.localizeArtifact(plugin.getArtifactId());
+      String pluginName = String.format("%s-%s-%s.jar",
+                                        plugin.getArtifactId().getScope().toString(),
+                                        plugin.getArtifactId().getName(),
+                                        plugin.getArtifactId().getVersion().toString());
+      RunJar.unJar(tempLocation, pluginsLocation.resolve(pluginName).toFile(), RunJar.MATCH_ANY);
+    }
+
+    // Fetching program.jar from app-fabric and expand it
+    Path programJarLocation = new File(WORKING_DIRECTORY).getAbsoluteFile().toPath();
+    File tempLocation = fetchArtifacts.localizeArtifact(spec.getArtifactId());
+    RunJar.unJar(tempLocation, programJarLocation.resolve(PROGRAM_JAR_EXPANDED_NAME).toFile(), RunJar.MATCH_ANY);
+    Files.copy(tempLocation.toPath(), programJarLocation.resolve(PROGRAM_JAR_NAME));
+
+    artifactFetcherService =
+      new ArtifactFetcherService(cConf, createBundle(new File(WORKING_DIRECTORY).getAbsoluteFile().toPath()));
+    artifactFetcherService.startAndWait();
+
+    SparkContainerLauncher.launch(delegateClass, delegateArgs.toArray(new String[delegateArgs.size()]));
+  }
+
+
+  private static ArtifactLocalizerClient createArtifactLocalizerClient(CConfiguration cConf, Configuration hConf)
+    throws Exception {
+    MasterEnvironment masterEnv = MasterEnvironments.create(cConf, "k8s");
+    if (masterEnv != null) {
```

if it's null we have a problem, maybe better to throw an exception